igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy

```diff
@@ -50,73 +64,72 @@ class DockerMonitor extends CommonPollingMonitor {
-                !(id in imageIds)
-            }.subscribe({ String imageId ->
-                log.info "Removing $imageId."
-                cache.remove(imageId)
-            }, {
-                log.error("Error: ${it.message}")
+            if (updateCache) {
+                delta.add(new ImageDelta(imageId: imageId, image: image))
             }
-            )
-            */
-
-            images.parallelStream().forEach({ TaggedImage image ->
-                def imageId = cache.makeKey(account, image.registry, image.repository, image.tag)
-                def updateCache = false
-
-                if (imageId in cachedImages) {
-                    if (trackDigests) {
-                        def lastDigest = cache.getLastDigest(image.account, image.registry, image.repository, image.tag)
-
-                        if (lastDigest != image.digest) {
-                            log.info("Updated tagged image: {}: {}. Digest changed from [$lastDigest] -> [$image.digest].", kv("account", image.account), kv("image", imageId))
-                            // If either is null, there was an error retrieving the manifest in this or the previous cache cycle.
-                            updateCache = image.digest != null && lastDigest != null
-                        }
-                    }
-                } else {
-                    log.info("New tagged image: {}: {}. Digest is now [$image.digest].", kv("account", image.account), kv("image", imageId))
-                    updateCache = true
-                }
+        })
 
-                if (updateCache) {
-                    postEvent(echoService, cachedImages, image, imageId)
-                    cache.setLastDigest(image.account, image.registry, image.repository, image.tag, image.digest)
-                }
-            })
-        } catch (Exception e) {
-            log.error("Failed to update account {}", kv("account", account), e)
-        }
+        return new DockerPollingDelta(items: delta, cachedImages: cachedImages)
+    }
+
+    /**
+     * IMPORTANT: We don't remove indexed images from igor due to the potential for
+     * incomplete reads from clouddriver or Redis.
+     */
+    @Override
+    void commitDelta(DockerPollingDelta delta) {
+        delta.items.parallelStream().forEach({ ImageDelta item ->
+            cache.setLastDigest(item.image.account, item.image.registry, item.image.repository, item.image.tag, item.image.digest)
+            postEvent(delta.cachedImages, item.image, item.imageId)
+        })
     }
 
     @Override
     String getName() {
         "dockerTagMonitor"
     }
 
-    void postEvent(EchoService echoService, List<String> cachedImagesForAccount, TaggedImage image, String imageId) {
+    void postEvent(List<String> cachedImagesForAccount, TaggedImage image, String imageId) {
+        if (!echoService.isPresent()) {
+            log.warn("Cannot send tagged image notification: Echo is not enabled")
```

Comment:
would it make sense to publish a metric when this happens?

Reply:
might not hurt